Switch to modern 'coverage' environment and run doctests on all versions; fix Travis builds

jamadden opened the pull request:
Based on #29 (because it has a fix needed to make the doctests run in pure environments.)

Drop Python 3.3.

Enable coveralls on Travis CI, and also fix using interpreters that aren't there in the new default 'trusty' image. Enable testing PyPy3 on Travis.

Coverage is not at 100% and in fact appears to have dropped (I show 92% in one run, vs the previous reported 97%), but that's because `coverage` and coveralls report files that aren't even imported (we have a lot of those in the 'tests' package)...nosetests didn't.

mgedmin commented on .coveragerc:
```diff
@@ -0,0 +1,12 @@
+[run]
+source = zope.security
```
Personally I think it's fine add `omit = */tests/*` to measure the coverage of the actual code, without having to pepper nocover pragmas for all the unused stubs/helpers in my tests.

jamadden replied:
I have held that opinion at times too. However, someone (I think maybe @tseaver) insisted that it was worth measuring test coverage too...and sure enough, not long after that I found a project with a whole bunch of tests that just weren't even being run. If test coverage was disabled, that wouldn't have happened.

I think the stubs issue isn't really that bad, at least not in my experience so far. It's actually a good thing to change a `def never_called(): pass` into a `def never_called(): raise AssertionError("why am I being called?")` to help make tests more explicit. And if there are bunches of unused/unimported files sitting around (like the case here)...well, why are they even there? It just adds extra overhead.

mgedmin commented on docs/api/checker.rst:
```diff
@@ -1021,4 +1024,4 @@ Always available:
    1
    >>> int(PBar.__class__ == type)
-   1
+   0
```
I admit that I don't understand this bit.

jamadden replied:
Previously, the doctests were only run on Python 3. The base class of `PBar` was defined using Python 2 syntax:
```python
class MyABC:
   __metaclass__ = abc.ABC
```
which didn't carry over to Python 3 (hence `class PBar(MyABC): pass; type(PBar)` *was* `type`. I fixed the test to always set up metaclasses correctly on both versions of Python, so they both properly report `PBar.__class__` as something other than `type`.

mgedmin replied:
Ah!  I saw the old tests using `__metaclass__ = ...` and other Python-2-nisms and assumed previously the tests were run only on 2.7.

Also, didn't [testenv:coverage] use basepython = python2.7?  I'm confused now.

jamadden replied:
Ah, but the coverage environment didn't run doctests. (*Only* the `docs` environment ran doctests, and it was---and still is---3.4.)

mgedmin replied:
I'm pretty sure all this `>>> int(foo == bar)` business is for compatibility with Python 2.2, where >>> True printed 1 and doctest couldn't cope.  Can we maybe get rid of the int()s and compare with True/False?

jamadden replied:
Yes, that can be done.

mgedmin replied:
Whoops, I don't know how I managed to confuse testenv:coverage with testenv:docs in my mind!

mgedmin commented on docs/api/permission.rst:
```diff
@@ -47,8 +47,9 @@ The :data:`zope.security.checker.CheckerPublic` permission always exists:
    >>> ids.sort()
-   >>> ids
-   ['x', 'y']
+   >>> for i in ids: print(i)
```
Might as well rewrite the above to `>>> ids = sorted(allPermissions(None))` while at it.

jamadden replied:
OK.

mgedmin commented on docs/api/permission.rst:
```diff
@@ -115,16 +116,18 @@ and its title is shortened:
 
-   >>> [term.title for term in vocab]
-   ['Public', 'x', 'y']
+   >>> for t in [term.title for term in vocab]: print(t)
```
This looks like a mechanical change.  I would prefer `>>> for term in vocab: print(term.title)`.

jamadden replied:
OK.

mgedmin commented on tox.ini:
```diff
@@ -18,34 +19,28 @@ setenv =
-    nosetests --with-xunit --with-xcoverage
+    coverage run -m zope.testrunner --test-path=src []
+    coverage run -a -m sphinx -b doctest -d {envdir}/.cache/doctrees docs {envdir}/.cache/doctest
+    coverage report --fail-under=92
```
Can't we put the --fail-under in `.coveragerc`?

Do we want Travis to fail if coverage decreases?  (I'm mildly in favour.)

jamadden replied:
coveralls.io has its own separate setting for failing if coverage decreases (and I currently have that on, set to the same thresholds). Since it doesn't get the source checkout (I think?) it can't read .coveragerc.

mgedmin commented on docs/api/checker.rst:
```diff
@@ -906,7 +906,11 @@ Classic Classes
+
+   >>> import sys
+   >>> len(C.__bases__) + (1 if sys.version_info[0] == 2 else 0)
    1
```
(I also do not understand this one.)

jamadden replied:
Above is `class C: pass`. On Python 2 that has no `__bases__`, but on Python 3, that's the same as `class C(object): pass`.

mgedmin replied:
Perhaps it would be clearer if you spelled it as

    >>> len(C.__bases__) == (0 if sys.version_info[0] == 2 else 1)
    True

jamadden replied:
How about:
```pycon
>>> isinstance(C, getattr(types, 'ClassType', type))
True
```

jamadden replied:
Oh, that doesn't work becasue `C` is proxied at this point.

jamadden commented:
>  I would be happier if this was rebased to drop the fix and the group_by fix then rebased on top of master when this is merged

Rebased (well, kind-of. I just dropped that commit from this PR, which I think accomplishes the same thing.)

jamadden commented:
Thank you! Now I'll rebase #29 on master.
